de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/FileSystemChangeListener.java

```diff
@@ -1,529 +1,870 @@
+
+            return true;
+        };
+
+        /*
+         * Calls the above defined contentIterator on the base directory and
+         * every contained resource.
+         * Directories are only stepped into if they match the above defined
+         * virtualFileFilter. This also applies to the base directory.
+         */
+        VfsUtilCore.iterateChildrenRecursively(oldFile, virtualFileFilter,
+            contentIterator);
+
+        while (!queuedDeletionActivities.isEmpty()) {
+            fireActivity(queuedDeletionActivities.pop());
         }
     }
 
-    @Override
-    public void fileCopied(
+    /**
+     * How the moved file is handled depends on whether the source and target
+     * directory are shared:
+     * <ul>
+     * <li>
+     * If only the source directory is shared, a deletion activity for the
+     * old file is created and dispatched. Subsequently removes any editor for
+     * the file from the editor pool and drops any annotation held for the file.
+     * </li>
+     * <li>
+     * If only the target directory is shared, a creation activity for the new
+     * file is created and dispatched. Subsequently adds an entry for the new
+     * file to the editor pool if it is currently open in an editor.
+     * </li>
+     * <li>
+     * If both the source and target directory are shared, a move activity for
+     * the file is created and dispatched. Subsequently updates the path of any
+     * editor for the file in the editor pool and updates the path for any held
+     * annotations for the file.
+     * </li>
+     * </ul>
+     * If the source directory is shared, an editor closed activity is also
+     * dispatched. If the target directory is shared, an editor opened activity
+     * is also dispatched. This is necessary as Intellij does not close and
+     * re-open the editors of moved files but rather changes the path of the
+     * editor internally, meaning the <code>UserEditorState</code> held by the
+     * other participants would not be updated correctly.
+     * <p></p>
+     * Renaming files is also handled as a move activity. This can be done
+     * with the optional parameter <code>newFileName</code>.
+     * <p></p>
+     * The optional parameter <code>newBaseName</code> can be used to handle
+     * cases where a parent file is renamed. The renamed directory has to be the
+     * directory located under <code>oldBaseParent</code> in the old path.
+     *
+     * @param oldFile       the file that is about to be moved
+     * @param oldBaseParent the old base parent of the file
+     * @param newBaseParent the new base parent of the file
+     * @param newBaseName   the new name for the base parent or null if it was
+     *                      not renamed
+     * @param newFileName   the new name for the file or null if it was not
+     *                      renamed
+     * @see #generateFolderMove(VirtualFile, VirtualFile, VirtualFile, String)
+     */
+    private void generateFileMove(
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Refactor this method to reduce its Cognitive Complexity from 20 to the 15 allowed. ([squid:S3776](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=307&issues=AWcYB0JQkEeD8QRk19vT&open=AWcYB0JQkEeD8QRk19vT)

[ISSUE_KEY]: AWcYB0JQkEeD8QRk19vT